### PR TITLE
Standardize MIOpen Documentation

### PR DESCRIPTION
**Pull request by samjwu**
Includes changes from Saad's RTD branch from PR https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2013

Use rocm-docs-core template for MIOpen Documentation

Rationale:
 - https://github.com/RadeonOpenCompute/rocm-docs-core#purpose

Note:
Links to old GitHub Pages to be updated in later PRs

### CODE REVIEW

**Comment by samjwu** [comment]
https://advanced-micro-devices-demo--2053.com.readthedocs.build/projects/MIOpen/en/2053/index.html

**Review by saadrahim — DISMISSED** [review]
Fixes my problems with doxygen rendering. Great starting point for our docs.

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# documentation artifacts
```
Comment (JehandadKhan) [inline .gitignore]: Please remove the `.gitignore` file from source control.

File: README.md [hunk 2]
```diff
@@ -8,6 +8,24 @@ MIOpen supports two programming models -
 1. [HIP](https://github.com/ROCm-Developer-Tools/HIP) (Primary Support).
 2. OpenCL.
 
+## Documentation
+
+For a detailed description of the **hipBLAS** library, its implemented routines, the installation process and user guide, see the [**hipBLAS** Documentation](https://hipblas.readthedocs.io/en/latest/).
```
Comment (JehandadKhan) [inline README.md]: Perhaps you mean `MIOpen` ?

File: docs/.doxygen/miopen.h
Comment (JehandadKhan) [inline docs/.doxygen/miopen.h:1]: Do we need to add the `doxygen` generated files to source control ?

File: docs/.sphinx/_toc.yml.in
Comment (JehandadKhan) [inline docs/.sphinx/_toc.yml.in:1]: Same as above

**Review by JehandadKhan — DISMISSED** [review]
The documentation CI step previously added by @saadrahim  is broken. Was that intended?

**Comment by samjwu** [comment]
> The documentation CI step previously added by @saadrahim is broken. Was that intended?

Yes, this was intended. The intent is to move away from GitHub pages and use Read the Docs for documentation to standardize ROCm documentation.

See: https://readthedocs.com/projects/advanced-micro-devices-miopen/builds/1381949/
and https://advanced-micro-devices-demo--2053.com.readthedocs.build/projects/MIOpen/en/2053/

File: README.md (hunk 2 above)
Comment (samjwu) [reply README.md]: Thanks for catching that. Fixed in [8a81f23](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2053/commits/8a81f23c396be26a19a859ae7c186400f390d639)

File: .gitignore (hunk 1 above)
Comment (samjwu) [reply .gitignore]: I can definitely do this, though I'm concerned it may cause some issues with accidentally adding documentation artifacts (eg: I forgot to add the copy of `miopen.h` to the .gitignore). Is there a reason we want to do this?

File: docs/.doxygen/miopen.h
Comment (samjwu) [reply docs/.doxygen/miopen.h:1]: It was not necessary. Thanks for catching that. Fixed in [8a81f23](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2053/commits/8a81f23c396be26a19a859ae7c186400f390d639)

File: docs/.sphinx/_toc.yml.in
Comment (samjwu) [reply docs/.sphinx/_toc.yml.in:1]: The _toc.yml.in file is necessary. It's a configuration used by Sphinx for the table of contents on the docs page: https://advanced-micro-devices-demo--2053.com.readthedocs.build/projects/MIOpen/en/2053/

File: .gitignore (hunk 1 above)
Comment (saadrahim) [reply .gitignore]: I'd like to add that we have added the .gitignore with similar contents to most other repositories. This was the first concern raised about a .gitignore. As a result, I would also like to understand your views on this.

File: docs/.sphinx/_toc.yml.in [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            - url: https://rocmsoftwareplatform.github.io/AMDMIGraphX/doc/html/
+              title: MIGraphX- Graph Optimization
+            - file: reference/computer_vision
+              subtrees:
+                - entries:
+                  - url: https://rocmdocs.amd.com/projects/MIVisionX/en/{branch}/
+                    title: MIVisionX
+                - entries:
+                  - url: https://rocmdocs.amd.com/projects/rocAL/en/{branch}/
+                    title: rocAL
```
Comment (pfultz2) [inline docs/.sphinx/_toc.yml.in]: Why is this file here? This seems to contain a lot of information not related to MIOpen.

File: docs/conf.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+#
+# This file only contains a selection of the most common options. For a full
+# list see the documentation:
+# https://www.sphinx-doc.org/en/master/usage/configuration.html
+
+import os
+from rocm_docs import ROCmDocs
+
+docs_core = ROCmDocs("MIOpen Documentation")
+docs_core.run_doxygen()
```
Comment (pfultz2) [inline docs/conf.py:10]: If you are running doxygen here then you should remove the doxygen run from the cmake.

File: docs/.doxygen/Doxyfile [hunk 5]
```diff
@@ -0,0 +1,2452 @@
...
+# the logo to the output directory.
+
+PROJECT_LOGO           = 
+
+# The OUTPUT_DIRECTORY tag is used to specify the (relative or absolute) path
+# into which the generated documentation will be written. If a relative path is
+# entered, it will be relative to the location where doxygen was started. If
+# left blank the current directory will be used.
+
+OUTPUT_DIRECTORY       = docBin
```
Comment (pfultz2) [inline docs/.doxygen/Doxyfile:61]: The output directory should be configured from a sphinx variable so rocm-cmake can put it in the build directory. 

Also, all paths should be all lowercase.

File: docs/conf.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+docs_core.setup()
+
+os.system("sed -e 's/MIOPEN_EXPORT //g' ../include/miopen/* > .doxygen/")
```
Comment (pfultz2) [inline docs/conf.py]: We shouldn't modify the source files when generating documentation.

File: docs/.sphinx/requirements.in [hunk 7]
```diff
@@ -0,0 +1 @@
+git+https://github.com/RadeonOpenCompute/rocm-docs-core.git
```
Comment (pfultz2) [inline docs/.sphinx/requirements.in:1]: This should pull from a specific tag or commit hash.

File: docs/.sphinx/_toc.yml.in (hunk 3 above)
Comment (pfultz2) [reply docs/.sphinx/_toc.yml.in]: If add or remove a project from rocm's documentation then we have to come to MIOpen and update this as well. This seems like the wrong way to handle this. Why cant this be stored in the rocm-docs-core instead?

File: docs/.sphinx/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+    #   sphinx-external-toc
+pyzmq==25.0.0
+    # via
+    #   ipykernel
+    #   jupyter-client
+requests==2.28.1
+    # via
+    #   pygithub
+    #   sphinx
+rocm-docs-core @ git+https://github.com/RadeonOpenCompute/rocm-docs-core.git
```
Comment (pfultz2) [inline docs/.sphinx/requirements.txt]: A specific commit hash should be used here.

File: docs/.sphinx/_toc.yml.in (hunk 3 above)
Comment (saadrahim) [reply docs/.sphinx/_toc.yml.in]: i think this is an accident. it will be fixed.

File: docs/.sphinx/requirements.txt (hunk 8 above)
Comment (saadrahim) [reply docs/.sphinx/requirements.txt]: Sam and I are preparing a bot account to make PRs to update this hash. We left it without a commit hash due to hourly changes we are making to rocm-docs-core. The intention is to pin it but in the interest of rapid development, we were a little lazy. If you want, we can pin it now or pin it for the rest of ROCm project simultaneosly via autogenerate PRs from the bot account.

**Comment by pfultz2** [comment]
Also does `make doc` still work to generate the documentation? Since you are standardizing the documentation, then why aren't replacing MIOpen's cmake documentation functions to use rocm-cmake's documentation functions as well?

https://radeonopencompute.github.io/rocm-cmake/doc/html/ROCMSphinxDoc.html

**Comment by saadrahim** [comment]
> Also does `make doc` still work to generate the documentation? Since you are standardizing the documentation, then why aren't replacing MIOpen's cmake documentation functions to use rocm-cmake's documentation functions as well?
> 
> https://radeonopencompute.github.io/rocm-cmake/doc/html/ROCMSphinxDoc.html

It is an iterative goal, we will approach the cmake builds for these projects soon. In the interest of timelines, the bare minimum was done. You can expect future PRs to standardize other parts of the builds. 

I suspect that we broke make doc, so I will take another look at it.

**Comment by pfultz2** [comment]
> It is an iterative goal, we will approach the cmake builds for these projects soon. In the interest of timelines, the bare minimum was done. You can expect future PRs to standardize other parts of the builds.

But we should at least make sure we didnt break the current `make doc`(and `make clean`) that is being used.

**Comment by saadrahim** [comment]
> > It is an iterative goal, we will approach the cmake builds for these projects soon. In the interest of timelines, the bare minimum was done. You can expect future PRs to standardize other parts of the builds.
> 
> But we should at least make sure we didnt break the current `make doc`(and `make clean`) that is being used.

As mentioned in the modified the earlier comment, I will fix the existing functionality.

File: docs/.sphinx/_toc.yml.in (hunk 3 above)
Comment (samjwu) [reply docs/.sphinx/_toc.yml.in]: Thanks for noting this. Fixed in [b42f3a6](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2053/commits/b42f3a65c3a49164055aff69aea6635e37158429)

File: docs/.doxygen/Doxyfile (hunk 5 above)
Comment (samjwu) [reply docs/.doxygen/Doxyfile:61]: This is used by rocm-docs-core, which expects the output to be in docBin/xml

Regarding path names, this should not appear in the repository as it is an intermediate used by the RTD build

**Comment by atamazov** [comment]
@samjwu Why we move .md files from doc/src to docs? This breaks links to .md files that scattered in comments over the repository.

Note that this "docs" is not in sync with existing tradition, e.g. we have src (not srcs), include (not includes) etc.

**Comment by samjwu** [comment]
> @samjwu Why we move .md files from doc/src to docs? This breaks links to .md files that scattered in comments over the repository.
> 
> Note that this "docs" is not in sync with existing tradition, e.g. we have src (not srcs), include (not includes) etc.

The main reason was to keep the same convention for documentation file structure of repositories (roc/hip BLAS, SOLVER, SPARSE, etc).

File: .gitignore (hunk 1 above)
Comment (JehandadKhan) [reply .gitignore]: In my opinion a `.gitignore` is developer specific since each developer uses a different development tool, the resulting artifacts vary. For instance I use VS code which create many file that need to be ignored specific to my dev env. Thus I can have a local `.gitignore` and add `.gitignore` to it followed by all the files I want to keep out of source control. This means that when I pull MIOpen into my local copy, there is no conflict. However, if we add a gitignore to the repo every time I do a pull there will be a merge conflict which would need to be resolved. 

If someone accidentally adds artifacts from their build, they will have to be reviewed and would be highlighted.

**Comment by JehandadKhan** [comment]
> > @samjwu Why we move .md files from doc/src to docs? This breaks links to .md files that scattered in comments over the repository.
> > Note that this "docs" is not in sync with existing tradition, e.g. we have src (not srcs), include (not includes) etc.
> 
> The main reason was to keep the same convention for documentation file structure of repositories (roc/hip BLAS, SOLVER, SPARSE, etc).

I believe moving the location is fine, if we can scan the code for links to files and update them as well.

File: .gitignore (hunk 1 above)
Comment (samjwu) [reply .gitignore]: Done in [remove .gitignore](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2053/commits/2b5535b784a9d0331f0dc8b861f0b96823d54c71)

File: docs/.sphinx/requirements.txt (hunk 8 above)
Comment (samjwu) [reply docs/.sphinx/requirements.txt]: Done in [pin rocm-docs-core](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2053/commits/3316da521c40ed56726d3e7106218db852d7ed97)

File: docs/conf.py (hunk 4 above)
Comment (samjwu) [reply docs/conf.py:10]: Removed cmake doc files in [de9b364](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2053/commits/de9b364d58ccbc9968a9fe79cbbcc4f763e873ed)

File: docs/conf.py (hunk 6 above)
Comment (samjwu) [reply docs/conf.py]: [81e377c](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2053/commits/81e377c569abcba1484353cce47c0018e67462bc)
Instead of changing the source file, this step will make a modified copy because doxygen cannot generate docs properly due to the MIOPEN_EXPORT prefix. The source files will remain the same.

File: README.md [hunk 9]
```diff
@@ -8,7 +8,24 @@ MIOpen supports two programming models -
...
 
+## Documentation
+
+For a detailed description of the **MIOpen** library see the [Documentation](https://rocmdocs.amd.com/projects/MIOpen/en/latest/).
+
+### How to build documentation
+
+Run the steps below to build documentation locally.
+
+```
```
Comment (saadrahim) [inline README.md:19]: If docs can build with cmake, we should include that as an alternate instruction.

File: docs/conf.py (hunk 6 above)
Comment (samjwu) [reply docs/conf.py]: Removed this line in [23e152a](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2053/commits/23e152af345139b45857c083504ec83490dff615)

File: docs/conf.py (hunk 4 above)
Comment (samjwu) [reply docs/conf.py:10]: Updated docs/CMakelists.txt to use rocm cmake [a48e0d3](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2053/commits/a48e0d3024988a8e018521aca397919b7d461274)

**Review by JehandadKhan — APPROVED** [review]
